### PR TITLE
Add sidecar webhook for injecting the vineyard container and Sidecar CRD for configuring the sidecar.

### PR DESCRIPTION
**Pull request by dashanji**
The pr does:
* Add sidecar webhook for injecting the vineyard container.
* Add Sidecar CRD and Controller for configuring the sidecar container.
* Add relevant e2e test.

### CODE REVIEW

File: k8s/apis/k8s/v1alpha1/sidecar_types.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	// metric configuration
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={}
+	Volume VolumeConfig `json:"volume,omitempty"`
+
+	// rpc service configuration
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={type: "ClusterIP", port: 9600, selector: "rpc.vineyardd.v6d.io/rpc=vineyard-rpc"}
+	Service ServiceConfig `json:"service,omitempty"`
+}
```
Comment (sighingnow) [inline k8s/apis/k8s/v1alpha1/sidecar_types.go:143]: Looks duplicate a lot with `vineyardd_types.go`...

**Review by sighingnow — COMMENTED** [review]
Could please finish the corresponding documentation as well?

File: k8s/apis/k8s/v1alpha1/sidecar_types.go (hunk 1 above)
Comment (dashanji) [reply k8s/apis/k8s/v1alpha1/sidecar_types.go:143]: No problem, I will add the doc in the next pr. Yeah, the Sidecar is much like Vineyardd, it is used to provide observability, showing users the detailed configuration of the vineyard after injection.

Comment (sighingnow) [reply k8s/apis/k8s/v1alpha1/sidecar_types.go:143]: Could you please try to make those duplicated structs into somewhere like `common.go` or `utils.go`? (I'm not sure whether kubebuilder can handle that.

Comment (sighingnow) [reply k8s/apis/k8s/v1alpha1/sidecar_types.go:143]: > I will add the doc in the next pr.

Why not in this pull request as well? I was imaging the feature and corresponding documentation should be treat as a whole.

File: k8s/config/crd/bases/k8s.v6d.io_sidecars.yaml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            type: string
+          kind:
+            type: string
+          metadata:
+            type: object
+          spec:
+            properties:
+              etcdCmd:
+                default: etcd
+                type: string
```
Comment (sighingnow) [inline k8s/config/crd/bases/k8s.v6d.io_sidecars.yaml:39]: I think the option `etcd_cmd` shouldn't be exposed to operator. On kubernetes, we luanch the etcd pods/deployments for vineyard and vineyard itself won't have the chance/possibility to find the etcd binary and luanch a etcd server in its container.

And indeed that won't work.

File: k8s/config/samples/k8s_v1alpha1_sidecar.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+apiVersion: k8s.v6d.io/v1alpha1
+kind: Sidecar
+metadata:
+  name: sidecar-sample
+spec:
+  #replicas: 1
+  image: test
```
Comment (sighingnow) [inline k8s/config/samples/k8s_v1alpha1_sidecar.yaml:7]: Looks wired. What is "test"?

File: k8s/pkg/operation/assembly.go [hunk 4]
```diff
@@ -154,12 +153,12 @@ func (ao *AssemblyOperation) buildLocalAssemblyJob(ctx context.Context, localObj
...
 			LocalAssemblyConfigTemplate.NodeName = localObject.Spec.Hostname
-			LocalAssemblyConfigTemplate.JobName = podLabels[schedulers.VineyardJobName]
+			LocalAssemblyConfigTemplate.JobName = podLabels["scheduling.k8s.v6d.io/job"]
 			LocalAssemblyConfigTemplate.TimeoutSeconds = timeout
 			if socket, err := ao.ResolveRequiredVineyarddSocket(
 				ctx,
-				podLabels[schedulers.VineyarddName],
-				podLabels[schedulers.VineyarddNamespace],
+				podLabels["scheduling.k8s.v6d.io/vineyardd"],
+				podLabels["scheduling.k8s.v6d.io/vineyardd-namespace"],
```
Comment (sighingnow) [inline k8s/pkg/operation/assembly.go:161]: Don't hard code strings. Use constant instead.

With constant variable, Go compiler will serve as a typo checker for you.

File: k8s/pkg/operation/assembly.go [hunk 5]
```diff
@@ -298,12 +297,12 @@ func (ao *AssemblyOperation) buildDistributedAssemblyJob(
...
 		DistributedAssemblyConfigTemplate.OldObjectToNewObject = str
-		DistributedAssemblyConfigTemplate.JobName = podLabels[schedulers.VineyardJobName]
+		DistributedAssemblyConfigTemplate.JobName = podLabels["scheduling.k8s.v6d.io/job"]
 		DistributedAssemblyConfigTemplate.TimeoutSeconds = timeout
 		if socket, err := ao.ResolveRequiredVineyarddSocket(
 			ctx,
-			podLabels[schedulers.VineyarddName],
-			podLabels[schedulers.VineyarddNamespace],
+			podLabels["scheduling.k8s.v6d.io/vineyardd"],
+			podLabels["scheduling.k8s.v6d.io/vineyardd-namespace"],
```
Comment (sighingnow) [inline k8s/pkg/operation/assembly.go:305]: Don't do that.

File: k8s/apis/k8s/v1alpha1/sidecar_types.go (hunk 1 above)
Comment (dashanji) [reply k8s/apis/k8s/v1alpha1/sidecar_types.go:144]: OK, If this doesn't burden your review, I'm willing to add it in this pr.

File: k8s/sidecar/sidecar_webhook.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+				},
+			},
+		},
+		Ports: []corev1.ContainerPort{
+			{
+				Name:          "vineyard-rpc",
+				ContainerPort: 9600,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		},
```
Comment (sighingnow) [inline k8s/sidecar/sidecar_webhook.go:201]: Maybe you could have a yaml as template for vineyardd container, then you could be relieve from the error-prone hacks in `BuildVineyardCommand`.

**Review by sighingnow — APPROVED** [review]

File: k8s/apis/k8s/v1alpha1/sidecar_types.go (hunk 1 above)
Comment (dashanji) [reply k8s/apis/k8s/v1alpha1/sidecar_types.go:144]: > Could you please try to make those duplicated structs into somewhere like `common.go` or `utils.go`? (I'm not sure whether kubebuilder can handle that.

Thanks for the advice, a possible way is to integrate some configurations of vineyard(image, command, etc) into a structure, store it in `vineyardd_types.go`, and then reuse this structure in `sidecar_types.go`, what do you think?

**Review by sighingnow — APPROVED** [review]
Basically LGTM. I have left some comments.

Thanks for the effort!

File: k8s/config/crd/bases/k8s.v6d.io_sidecars.yaml (hunk 2 above)
Comment (dashanji) [reply k8s/config/crd/bases/k8s.v6d.io_sidecars.yaml:39]: Make sense, I'll update it later.

File: k8s/config/samples/k8s_v1alpha1_sidecar.yaml (hunk 3 above)
Comment (dashanji) [reply k8s/config/samples/k8s_v1alpha1_sidecar.yaml:7]: Ummm, sorry about it. It's only for testing, I will add a complete cr next.

File: k8s/pkg/operation/assembly.go (hunk 4 above)
Comment (dashanji) [reply k8s/pkg/operation/assembly.go:161]: Thanks for the catch. In fact, this change is to solve the problem of mutual import of golang packages, I will put all global variables into a golang package to solve it.

File: k8s/sidecar/sidecar_webhook.go [hunk 7]
```diff
@@ -80,204 +86,107 @@ func (r *SidecarInjector) Handle(ctx context.Context, req admission.Request) adm
...
+			// use default configurations
 			sidecar.Spec.Selector = keys[0] + "=" + labels[keys[0]]
-			if err := r.Client.Create(ctx, &sidecar); err != nil {
+			sidecar.Spec.VineyardConfig.EtcdEndpoint = "http://etcd-for-vineyard:2379"
+			sidecar.Spec.VineyardConfig.EtcdPrefix = "/vineyard"
+			sidecar.Spec.VineyardConfig.Image = "vineyardcloudnative/vineyardd:latest"
+			sidecar.Spec.VineyardConfig.ImagePullPolicy = "IfNotPresent"
+			sidecar.Spec.VineyardConfig.Socket = "/var/run/vineyard.sock"
+			sidecar.Spec.VineyardConfig.Size = "256Mi"
+			sidecar.Spec.VineyardConfig.StreamThreshold = 80
```
Comment (sighingnow) [inline k8s/sidecar/sidecar_webhook.go:105]: Why do we need these hard-coded things here?

I think they should be specified in the sidecar's yaml?

**Review by sighingnow — APPROVED** [review]

File: k8s/sidecar/sidecar_webhook.go (hunk 7 above)
Comment (sighingnow) [reply k8s/sidecar/sidecar_webhook.go:105]: Can that be fixed? It not, maybe moving those constants to `_types.go` is a better solution.

Comment (dashanji) [reply k8s/sidecar/sidecar_webhook.go]: Fixed. It's actually a bug, as the default value is not the same between the nested structure and its' sub-fields.

**Pull request merged**